### PR TITLE
Explore changing /firefox/all/ to be less process intensive.

### LINKED ISSUE #9845: Improve firefox/all/ page performance
## Description

See: https://bugzilla.mozilla.org/show_bug.cgi?id=1683671

TL:DR: The HTML payload for the /firefox/all/ page is 1.5MB, which can be a performance issue. The time it takes for the back-end to return the document also seems higher than other pages, likely because of all the links it needs to render in the hidden tables. We often see test failures on this page because the page was too slow to render.

We should try and improve both the front-end and back-end performance of this page. Hopefully there's a way without requiring a significant redesign / rebuild.

### LINKED ISSUE #14673: Add Microsoft Store as installer option on firefox/all
## Description

Add links to the Microsoft Store versions of Firefox on /firefox/all.

- [x] Add links to the Microsoft Store in the list of installers for Release and Beta
- [x] Auto fill the language as "multiple languages" like we do for iOS and Android
- [x] Display a link to the Microsoft store in step 4 instead of a download button 

Firefox: https://www.microsoft.com/store/apps/9NZVDKPMR9RD
Firefox Beta: https://www.microsoft.com/store/apps/9NZW26FRNDLN
Firefox Nightly: coming soon.

- [x] Add this text to the installer help modal:

> Microsoft Store
> This option will open a Firefox product page in the Microsoft Store web portal. If you are running Windows 10 or Windows 11, you will have the option to open the Microsoft Store from this page and install Firefox. If you install from the Microsoft Store, updates to Firefox will also be handled by the Microsoft Store. There are minor differences in Firefox behaviors and capabilities when installed from the Microsoft Store, but for most users these differences are not noticeable.
> 

- [x]  Make sure strings are FTL'd

### PR DESCRIPTION
**Pull request by robhudson**
## One-line summary

This branch is an effort to eventually replace `/firefox/all/` to avoid a very heavy paylaod to the browser by breaking the choices into separate pages rather than a single page.

- [x] I used an AI to write some of this code (sh: I used Co-pilot while writing the tests)

## Changes

- [x] Improve wording to better explain this is a multi-step selection, not just choosing a language
- [x] Re-write CSS for new page structure, lean up old CSS classes and CSS file
- [x] Remove JS that supported old single page structure
- [x] Remove JS tests
- [x] Re-write integration tests
- [x] Add a cancel / back button for each step
- [x] Add a single page for both mobile platforms to simplify mobile download process
- [x] Remove template all-unified.html and linked css/js/tests
- [x] Display system requirements, release notes and support links at download step
- [x] Update [URL helper](https://github.com/mozilla/bedrock/blob/d83577cc7736bb9cab5de85ea3cb45a215fd5d9a/bedrock/firefox/templatetags/helpers.py#L298)
- [x] Update links with anchors to #products-*

## Issue / Bugzilla link

Fix #9845 
Fix #14673 

## Testing

https://www-demo2.allizom.org/en-US/firefox/all/

- [x] All the channel/os combos from the current /all page have equivalents in these pages
- [x] The download buttons download what they say they should
- [x] The Micosoft Store button displays for store downloads and is localized when appropriate
- [x] Do we have good test coverage?

Analytics:
- [x] Desktop downloads have stub attribution & GA4
- [x] Mobile download buttons have GA4

Mobile:
- [x] Test app & play store links on an actual mobile device.

Localization:
- [x] No untranslated strings
- [x] No unused translations (they should be deleted if we're not using them)
- [x] RTL support

QR codes go to expected App/Play store:
- [x] Mobile
- [x] Android Beta
- [x] Android Nightly
- [x] Scan one with a phone to make sure it has enough size/whitespace to work

### CODE REVIEW

**Comment by stephaniehobson** [comment]
I've started but I'm not finished and I don't have time to leave notes for someone else to pick it up :/ Fortunately, there's no rush.

**Comment by stephaniehobson** [comment]
I pushed my most recent changes.

- Links that need changing are `href="TODO"`
  - There should be 3 "back buttons" plus the download button
- The page that needs fixing is linked on the first page under Browser > Mobile > Firefox

**Comment by stephaniehobson** [comment]
@alexgibson This will change /all to work with a folder structure instead of anchors. Do you think that is likely to break anything?

**Comment by alexgibson** [comment]
> @alexgibson This will change /all to work with a folder structure instead of anchors. Do you think that is likely to break anything?

There are a few places in bedrock where we point to the old anchors I think, which should be easy enough to update. I have no idea about external sites though. Maybe referrer in GA might point us to any other sites that link to it?

**Comment by stephaniehobson** [comment]
I added some tests 🎉  And, like any good tests, they have discovered some bugs 🐞

File: bedrock/firefox/templates/firefox/includes/download-unsupported.html [hunk 1]
```diff
@@ -30,7 +30,7 @@
     </p>
 
     <small class="fx-unsupported-message-all-link">
-      <a href="{{ url('firefox.all') }}#product-desktop-esr">
+      <a href="{{ firefox_url('desktop', 'all', 'esr') }}>
```
Comment (janbrasna) [inline bedrock/firefox/templates/firefox/includes/download-unsupported.html:33]: ah found it:

```suggestion
      <a href="{{ firefox_url('desktop', 'all', 'esr') }}">
```
<details><summary>sshot ↓</summary>

Screen Shot 2024-07-18 at 15 39 15

</details>

File: bedrock/firefox/templates/firefox/all/platform.html [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        <li>
+          <h4 class="c-help-title">{{ ftl('firefox-all-msi-installers') }}</h4>
+          <p class="c-help-desc">
+            {{ ftl('firefox-all-windows-installers-for') }}
+          </p>
+        </li>
+        <li>
+          <h4 class="c-help-title">{{ ftl('firefox-all-arm64-installers') }}</h4>
+          <p class="c-help-desc">
+            {{ ftl('firefox-all-arm64-installers-optimized') }}
```
Comment (janbrasna) [inline bedrock/firefox/templates/firefox/all/platform.html:61]: There is a new string reflecting linux aarch64 addition:

```suggestion
            {{ ftl('firefox-all-arm64-installers-optimized-v2') }}
```

(This one is being currently removed as obsolete for some months already.)

**Review by janbrasna — CHANGES_REQUESTED** [review]
Now with the recent rebase even AArch64 Linux Nightlies started working, nice job! 🚀 

Right now I'm debugging general layout issues with this branch, pretty site-wide, and can't seem to be finding any culprits in the added styles though:/ I'm constantly having some weird layout issues for this PR when spoofing UAs (or it might be related to OSX versions being frozen in UAstring to 10_15_7 that's also the current support fringe?) like content blinking or disappearing when scrolling in webkit (on unrelated pages like fx/new) or text being dramatically smaller than what it should be in gecko etc. The most visible is ie the CTA alignment or paragraph sizing, anywhere on the site, not just under /firefox or /firefox/all etc.:

Screen Shot 2024-07-18 at 15 51 20

Screen Shot 2024-07-18 at 15 50 39

**Comment by stephaniehobson** [comment]
@janbrasna nice find on the missing `"` 😵

**Comment by stephaniehobson** [comment]
Ready for review 🎉

**Comment by janbrasna** [comment]
Oh I see, I was too nosy/early 😇 _(but just love this PR and kept an eye on it for a while;)…)_

One super-minor layout issue I've encountered is at some widths, the chevrons don't play well with longer strings:

https://github.com/user-attachments/assets/da6b358a-3117-4eaf-9713-1b8579b1be98

(but I guess this Is so minor it can be sorted out later not blocking this going to prod…)

- - - - -

The "Release notes" link for Android beta https://www-demo2.allizom.org/en-US/firefox/android/beta/notes/ leads to a version from 2020, that's probably when they stopped publishing them for beta? I can also open an issue for this separately, as this looks like a more meta issue (i.e. should the notes even redirect there, to some v68.x?)

- - - - -

I can make it a sad 500 if I feel fancy and put the `/win-store/` after any other product, like https://www-demo2.allizom.org/firefox/all/desktop-nightly/win-store/ (which is happily displayed instead of barking at me) and then clicking any language, obviously.

- - - - -

>  nice find on the missing `"` 😵

Ah that wasn't really me, after weeks of frustration why the layout order doesn't make sense I ran it through all the validators, parsers and formatters I could think of in desperation, eventually pointing out unexpected tokens leading me to areas with something not quite closed or quoted, and the rest was easy;)

FYI the nesting fix also resolved the violent flickering in Safari when scrolling as shown on one of the screen captures, so good it was the same cause… 🤞

>  will change /all to work with a folder structure instead of anchors. Do you think that is likely to break anything?

Good thing is it's from anchors to folders, so any outdated external links to anchors would just land the users at a root "step 0" initial state, still being able to make the selection manually and finish the task.

(I can imagine this may get a lightweight progressive JS treatment at a later stage that would only return the options fragments and populate them in-place, instead of full page reloads, that if making use of History API can also be translating the old fragment paths to the new folder structure URLs "as a rewrite"?)

BTW the Playwright tests are currently commented out as they liked to time out on the old /all — this new structure now enables writing new tests for it in Playwright too.

**Comment by robhudson** [comment]
If we need to we can rebase this. I don't have any local changes so it's fine to rebase and force push.

File: tests/functional/firefox/test_all.py [hunk 3]
```diff
@@ -6,89 +6,276 @@
...
+    assert modal.is_displayed
+
+
+@pytest.mark.smoke
+@pytest.mark.nondestructive
+def test_open_installer_help_modal(base_url, selenium):
+    slug = "/desktop-release/"
+    page = FirefoxAllPage(selenium, base_url, slug=slug).open()
+    modal = page.open_help_modal("icon-installer-help")
+    assert modal.is_displayed
```
Comment (robhudson) [inline tests/functional/firefox/test_all.py:281]: 🎉 nice tests

**Review by robhudson — COMMENTED** [review]
I can't approve my own PR, but I read through things and it all is looking good to me from the code.

**Review by janbrasna — CHANGES_REQUESTED** [review]
I'm afraid this misses the functionality to choose from several ESR versions.

The current https://www.mozilla.org/firefox/all/#product-desktop-esr offers the choice of esr-latest vs. esr-next — and most importantly: it's the only place to download it, all the ESR CTAs around the web only link to esr-latest (=115.x), users looking to download esr-next (=128.x) can only find it in /all — and I'm having trouble locating a step where I could make such a choice here _(admittedly now only testing demo2, but that's not too old iirc)_

**Comment by stephaniehobson** [comment]
@janbrasna Yeah, we missed that. I poked at it for a couple hours but didn't make any progress adding the version. I'll enlist someone else's help (probably not Rob since he's away for 3 weeks).

**Comment by stephaniehobson** [comment]
Added support for multiple ESR versions with Rob's help.

**Review by janbrasna — CHANGES_REQUESTED** [review]
I wish I could have helped more, fixing things instead of just breaking them again;)

The tests (with 10yo+ product-details snapshot and archaic version numbers;D) now pass, the href is no longer empty, only the version numbers don't match as the _demo2 product-details_ is couple of weeks old according to `healhz-cron` so it still says 128.0esr while the -next-latest already supplies 128.1.0esr _(any way to nudge the demos to refresh the external data?)_, and ofc stage bouncer sends outdated files for correct ids anyways;)

I thought the empty href started appearing when the version moved to 128.1.0esr, but then some of the envs still had 128.0esr yet the href stayed empty so it was unrelated and the last fix resolved that. I also tried to find an unreleased locale but it seems all languages are available right away(?) so I may need to dig deeper to find a release that's not existent to test it;]

Some observations for the recent updates:

File: bedrock/firefox/templates/firefox/all/base.html [hunk 4]
```diff
@@ -17,6 +17,11 @@
  {{ css_bundle('firefox_all') }}
 {% endblock %}
 
+{# do not index child pages, we prefer people visit more user friendly pages from search results #}
+{% block canonical_urls %}
+{% if product %}<meta name="robots" content="noindex,follow">{% endif %}
+{% endblock %}
+
```
[the remark this answers is not in the document]
Comment (stephaniehobson) [reply bedrock/firefox/templates/firefox/all/base.html]: The suggestion didn't work unfortunately, I had to come at it with a `super()` like you suggested initially.

File: bedrock/firefox/views.py [hunk 5]
```diff
@@ -320,11 +320,18 @@ class MobileRelease:
...
-        if not download_url and esr_next_version:
-            # TKTK: this probably needs some kind of saftey around the locale availability
-            context["download_esr_next_url"] = list(
-                filter(lambda b: b["locale"] == locale, firefox_desktop.get_filtered_full_builds(channel_esr_next, esr_next_version))
-            )[0]["platforms"][platform]["download_url"]
+        if esr_next_version:
+            try:
+                download_esr_next_url = list(filter(lambda b: b["locale"] == locale, firefox_desktop.get_filtered_full_builds("esr_next")))[0][
+                    "platforms"
+                ][platform]["download_url"]
```
[the remark this answers is not in the document]
Comment (stephaniehobson) [reply bedrock/firefox/views.py:327]: I have mobile failing for me now too. A bunch of the edge cases are unhappy :( 

Also, none of the tests caught that so.... more tests!

[the remark this answers is not in the document]
Comment (stephaniehobson) [reply bedrock/firefox/views.py:327]: I think I fixed it. 🤞

**Review by pmac — APPROVED** [review]
I love this. Gives great URLs for direct linking as well. Great work 💯

File: l10n/en/firefox/all.ftl [hunk 6]
```diff
@@ -4,31 +4,58 @@
 
 ### URL: https://www-dev.allizom.org/firefox/all/
 
-firefox-all-check-the-system-requirements = Check the system requirements
-firefox-all-release-notes = Release notes
-firefox-all-source-code = Source code
-firefox-all-need-help = Need help?
-firefox-all-which-browser-would = Which browser would you like to download?
+# HTML page title. Replace "English (US)" with your local language.
+firefox-all-download-the-firefox = Download the { -brand-name-firefox-browser } in English (US) and more than 90 other languages
```
Comment (alexgibson) [inline l10n/en/firefox/all.ftl:8]: **nit:** should this be "Firefox browser"?

```suggestion
firefox-all-download-the-firefox = Download the { -brand-name-firefox } browser in English (US) and more than 90 other languages
```

File: l10n/en/firefox/all.ftl [hunk 7]
```diff
@@ -4,31 +4,58 @@
...
+
+# HTML page description, also used as the introductory text.
+firefox-all-everyone-deserves-access = Everyone deserves access to the internet — your language should never be a barrier. That’s why — with the help of dedicated volunteers around the world — we make the { -brand-name-firefox-browser } available in more than 90 languages.
```
Comment (alexgibson) [inline l10n/en/firefox/all.ftl:11]: **nit:** avoid older product brand term in new strings

```suggestion
firefox-all-everyone-deserves-access = Everyone deserves access to the internet — your language should never be a barrier. That’s why — with the help of dedicated volunteers around the world — we make the { -brand-name-firefox } browser available in more than 90 languages.
```

File: bedrock/firefox/templates/firefox/all/product.html [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  {% if not product %}
+  <a id="icon-browser-help" href="#browser-help" class="c-button-help" title="{{ ftl('firefox-all-learn-about-firefox') }}">
+    {{ ftl('firefox-all-get-help') }}
+  </a>
+  {% endif %}
+</h2>
+
+{% if not product %}
+  <div class="c-step-contents">
+    <h3 class="c-product-title"><img alt="" role="presentation" src="{{ static('protocol/img/icons/desktop.svg') }}"> Desktop</h3>
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/all/product.html:26]: **issue:** should "Desktop" be translated?

File: bedrock/firefox/templates/firefox/all/product.html [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    <ul class="c-product-list qa-desktop-list">
+      <li class="release"><a href="{{ url('firefox.all.platforms', product_slug='desktop-release') }}">Firefox</a> <strong>(Recommended)</strong></li>
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/all/product.html:28]: **issue:** should "(Recommended)" be translated?

File: bedrock/firefox/templates/firefox/all/product.html [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+      {% for p in products %}
+        {% if p.slug.startswith('desktop') and not p.slug.endswith('release') %}
+          <li class="{{ p.slug }}"><a href="{{ url('firefox.all.platforms', product_slug=p.slug) }}">{{ p.name }}</a></li>
+        {% endif %}
+      {% endfor %}
+    </ul>
+
+    <h3 class="c-product-title"><img alt="" role="presentation" src="{{ static('protocol/img/icons/mobile.svg') }}"> Mobile</h3>
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/all/product.html:36]: **issue:** should "Mobile" be translated?

File: bedrock/firefox/templates/firefox/all/download.html [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+#}
+
+{% from "macros.html" import google_play_button, apple_app_store_button, ms_store_button with context %}
+
+{% set icon_download ='<span class="mzp-c-button-icon-end"><svg xmlns="http://www.w3.org/2000/svg" xml:space="preserve" viewBox="0 0 16 16" width="16" height="16" fill="currentColor"><path d="M8 13c.2 0 .4-.1.5-.2l4.4-4.4-1.1-1.1-3.1 3.1V1H7.2v9.4L4.1 7.3l-1 1.1 4.4 4.4c.1.1.3.2.5.2z" /><path d="M13.5 12v2c0 .3-.2.5-.5.5H3c-.3 0-.5-.2-.5-.5v-2H1v2c0 1.1.9 2 2 2h10c1.1 0 2-.9 2-2v-2h-1.5z" /></svg></span>' %}
+
+{# Show download #}
+
+<h2 class="c-step-name{% if not (product and platform and locale) %} t-step-disabled{% endif %}">
+  4. {{ ftl('firefox-all-download') }}
```
Comment (alexgibson) [inline bedrock/firefox/templates/firefox/all/download.html:14]: **question:** this might be one for the L10n team, but might some locales prefer the step numbers here to be part of the localized string?

**Review by alexgibson — CHANGES_REQUESTED** [review]
I took a first-pass and spotted a few issues but nothing all hopefully minor / easy enough to fix. Excited to see this one land after all this time!

File: bedrock/firefox/templates/firefox/all/download.html (hunk 11 above)
Comment (janbrasna) [reply bedrock/firefox/templates/firefox/all/download.html:14]: Oh that's a good note — if they were OL it won't be an issue (UA handles that with lang and css defaults), but truth is when they mimic an ordinal sequence they may need to be localised along with the text in languages that don't use `latn` numbers? (as `arab` `beng` `deva`…)

File: bedrock/firefox/urls.py [hunk 12]
```diff
@@ -36,6 +36,9 @@
         ),
     ),
     path("firefox/all/", views.firefox_all, name="firefox.all"),
+    path("firefox/all/<slug:product_slug>/", views.firefox_all, name="firefox.all.platforms"),
+    path("firefox/all/<slug:product_slug>/<str:platform>/", views.firefox_all, name="firefox.all.locales"),
```
[the remark this answers is not in the document]
Comment (janbrasna) [reply bedrock/firefox/urls.py:40]: The same passes for any value in win-store:
https://www-demo2.allizom.org/en-US/firefox/all/desktop-release/win-store/aa-ZZ/

OTOH here are some 500s on bogus keys:
https://www-demo2.allizom.org/en-US/firefox/all/desktop-beta/linux64/aa-ZZ/
https://www-demo2.allizom.org/en-US/firefox/all/desktop-beta/linux00/
https://www-demo2.allizom.org/en-US/firefox/all/desktop-bogus/

also using nightly-only language with other channels:
https://www-demo2.allizom.org/en-US/firefox/all/desktop-release/linux64/hye/

and passing win-store to other products:
https://www-demo2.allizom.org/en-US/firefox/all/mobile-beta/win-store/
https://www-demo2.allizom.org/en-US/firefox/all/desktop-developer/win-store/ (this one won't fail, but all the language links would 500 from here on…)

File: tests/functional/firefox/test_all.py [hunk 13]
```diff
@@ -6,89 +6,276 @@
...
+
+
+@pytest.mark.smoke
+@pytest.mark.nondestructive
+@pytest.mark.parametrize("slug, count", CHANNEL_PLATFORM_COUNT)
+def test_desktop_platform_list(slug, count, base_url, selenium):
+    slug = f"{slug}/"
+    page = FirefoxAllPage(selenium, base_url, slug=slug).open()
+    list_length = page.platform_list_length
+    assert list_length == count
```
[the remark this answers is not in the document]
Comment (robhudson) [reply tests/functional/firefox/test_all.py:40]: We have similar tests in the unit tests already too that could be fleshed out a bit more to cover what these are testing. I'll update them.

File: bedrock/firefox/templates/firefox/all/download.html [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+      {# If there are multiple ESR's available, show next as well. #}
+      {% if is_esr_next_version %}
+        <p class="c-step-download">
+          <a href="{{ download_esr_next_url }}" class="mzp-c-button mzp-t-product mzp-t-xl c-download-button" data-download-version="{{ platform.slug }}" data-cta-text="Download Now">
+            {{ ftl('firefox-all-download-esr-version', esr_version=desktop_esr_next_version) }} {{ icon_download|safe }}
+          </a>
+        </p>
+        <p>{{ ftl('firefox-all-esr-we-are-in-transition', esr_next_major=desktop_esr_next_version.split('.')[0], esr_current_major=desktop_esr_latest_version.split('.')[0] ) }}</p>
+        <p>{{ ftl('firefox-all-esr-when-there-is') }}</p>
+        <p>{{ ftl('firefox-all-esr-if-you-choose') }}</p>
```
[the remark this answers is not in the document]
Comment (stephaniehobson) [reply bedrock/firefox/templates/firefox/all/download.html:52]: ✅

File: bedrock/firefox/templates/firefox/all/download.html [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+      <p class="c-step-download">
+        <a href="{{ download_url }}" class="mzp-c-button mzp-t-product mzp-t-xl c-download-button" data-download-version="{{ platform.slug }}" data-cta-text="Download Now">
+          {{ ftl('download-button-download-now') }} {{ icon_download|safe }}
+        </a>
+      </p>
+      {% endif %}
+    {% else %}
+      {# All other desktop releases #}
+      <p class="c-step-download">
+        <a href="{{ download_url }}" class="mzp-c-button mzp-t-product mzp-t-xl c-download-button" data-download-version="{{ platform.slug }}" data-cta-text="Download Now">
```
[the remark this answers is not in the document]
Comment (stephaniehobson) [reply bedrock/firefox/templates/firefox/all/download.html]: ✅ 

This should be fixed now. The change to stub-attribution.js might be over-kill but it's actually more performant than the more restricted one.

File: l10n/en/firefox/all.ftl (hunk 6 above)
Comment (stephaniehobson) [reply l10n/en/firefox/all.ftl:8]: These show up as altered because I moved them up to the top of the file, they're not new strings.

That said, I might as well fix them while I'm in here.

File: l10n/en/firefox/all.ftl (hunk 7 above)
Comment (stephaniehobson) [reply l10n/en/firefox/all.ftl:11]: These show up as altered because I moved them up to the top of the file, they're not new strings.

That said, I might as well fix them while I'm in here.

File: bedrock/firefox/templates/firefox/all/product.html (hunk 10 above)
Comment (stephaniehobson) [reply bedrock/firefox/templates/firefox/all/product.html]: ✅

File: bedrock/firefox/templates/firefox/all/product.html (hunk 9 above)
Comment (stephaniehobson) [reply bedrock/firefox/templates/firefox/all/product.html]: ✅

File: bedrock/firefox/templates/firefox/all/download.html (hunk 11 above)
Comment (stephaniehobson) [reply bedrock/firefox/templates/firefox/all/download.html]: ✅  Moved the numbers into the strings.

File: bedrock/firefox/urls.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (robhudson) [reply bedrock/firefox/urls.py:40]: Yeah we need some validation on the platform, os, and lang being based in here. I can work on that. Thanks for catching that.

File: bedrock/firefox/tests/test_firefox_all.py [hunk 16]
```diff
@@ -0,0 +1,315 @@
...
+    assert "https://support.mozilla.org/kb/install-firefox-linux" in linux_link.attr("href")
+
+
+@pytest.mark.parametrize("os, lang", OS_LANG_PAIRS)
+def test_firefox_esr(client, os, lang):
+    resp = client.get(reverse("firefox.all.download", kwargs={"product_slug": "desktop-esr", "platform": os, "locale": lang}))
+    doc = pq(resp.content)
+
+    link = doc(".c-download-button")
+    assert len(link) == 1
```
Comment (janbrasna) [inline bedrock/firefox/tests/test_firefox_all.py:244]: I know these tests run against a static copy of product-details (not sure if there's both esr-lastest and esr-next pointing to different versions at that v50.x point in time), but IRL there may be two instances of the button…

**Review by janbrasna — COMMENTED** [review]
Only observations / naïve questions:

File: bedrock/firefox/tests/test_firefox_all.py [hunk 17]
```diff
@@ -0,0 +1,315 @@
...
+    download_url = link.attr("href")
+    if "msi" in os:
+        product = "firefox-nightly-msi-latest-l10n-ssl"
+        os = os.replace("-msi", "")
+    else:
+        product = "firefox-nightly-latest-l10n-ssl"
+    if lang == "en-US":
+        # en-us downloads don't get the l10n releases
+        product = product.replace("-l10n", "")
```
[the remark this answers is not in the document]
Comment (robhudson) [reply bedrock/firefox/tests/test_firefox_all.py:231]: Thanks! Updated in next commit.

File: bedrock/firefox/tests/test_firefox_all.py (hunk 16 above)
Comment (robhudson) [reply bedrock/firefox/tests/test_firefox_all.py:244]: Ah yes, we should add a test for that. I just need to figure out how to add the `esr-next` in the test. I'll take a look.

Comment (robhudson) [reply bedrock/firefox/tests/test_firefox_all.py:238]: There's now a test for esr-next and 2 links. Patching the product details was a bit ugly.

File: bedrock/firefox/urls.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (robhudson) [reply bedrock/firefox/urls.py:40]: I've added 404 checks for all of the above now in the latest commits.

**Review by alexgibson — APPROVED** [review]
Updates look good to me, great work @robhudson @stephaniehobson!

The only remaining thing that's missing from my initial pass is RTL styles for the product icons. Once that's done I think this is good to merge. 

r+wc
